src/lem/constrainer.rs

```diff
@@ -81,103 +150,565 @@ impl LEM {
                 .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path")),
         }
     }
 
+    fn z_ptr_from_valuation<F: LurkField>(
+        path_kind: &PathKind,
+        valuation: &Valuation<F>,
+        name: &String,
+        store: &mut Store<F>,
+    ) -> Result<ZPtr<F>> {
+        if Self::on_concrete_path(path_kind)? {
+            let Some(ptr) = valuation.ptrs.get(name) else {
+                bail!("Couldn't retrieve {} from valuation", name);
+            };
+            store.hydrate_ptr(ptr)
+        } else {
+            Ok(ZPtr::dummy())
+        }
+    }
+
     fn allocate_and_inputize_input<'a, F: LurkField, CS: ConstraintSystem<F>>(
         &'a self,
         cs: &mut CS,
         store: &mut Store<F>,
-        witness: &Witness<F>,
+        valuation: &Valuation<F>,
         alloc_ptrs: &mut HashMap<&'a String, AllocatedPtr<F>>,
         idx: usize,
     ) -> Result<()> {
-        if alloc_ptrs.contains_key(&self.input[idx]) {
-            bail!("{} already allocated", &self.input[idx]);
-        }
         let alloc_ptr = Self::allocate_ptr(
-            &mut cs.namespace(|| format!("allocate input {}", &self.input[idx])),
-            &store.hydrate_ptr(&witness.input[idx])?,
+            cs,
+            &store.hydrate_ptr(&valuation.input[idx])?,
             &self.input[idx],
+            alloc_ptrs,
         )?;
         alloc_ptrs.insert(&self.input[idx], alloc_ptr.clone());
         Self::inputize_ptr(cs, &alloc_ptr, &self.input[idx])?;
         Ok(())
     }
 
-    /// Create R1CS constraints for LEM given an evaluation witness.
+    fn stack_hash_slots<F: LurkField>(
+        path_kind: &PathKind,
+        hash_slots: &mut HashSlots<F>,
+        slots: &SlotsIndices,
+        hash: MetaPtr,
+        alloc_arity: AllocHashPreimage<F>,
+        tag: Option<Tag>,
+    ) -> Result<()> {
+        let is_concrete_path = Self::on_concrete_path(path_kind)?;
+        match alloc_arity {
+            AllocHashPreimage::A2(i0, i1) => {
+                if let PathKind::MaybeConcrete(concrete_path) = path_kind {
+                    // if concrete_path is true, push to slots
+                    if is_concrete_path {
+                        hash_slots.hash2_alloc.push((slots.hash2_idx, i0, i1));
+                        // only once per path
+                    }
+                    // concrete path implies alloc_tgt has the same value as in the current slot
+                    hash_slots.hash2_stacks.implies_stack.push((
+                        concrete_path.clone(),
+                        slots.hash2_idx,
+                        hash,
+                        tag,
+                    ));
+                // many
+                } else {
+                    hash_slots
+                        .hash2_stacks
+                        .enforce_stack
+                        .push((slots.hash2_idx, hash, tag));
+                    hash_slots.hash2_alloc.push((slots.hash2_idx, i0, i1))
```

Comment:
There is so much repetitive code here — between the different arities and the different 'stack types', that what should be a relatively simple meaning is very difficult to extract and/or verify. For the former, hopefully some combination of macros and a more a generic representation will help. For the latter, I think the idea of maintaining multiple stacks/queues makes this a lot more complex than necessary. 

Instead, what if you had an enum that distinguished between conditional ('implies') and unconditional ('enforce') constraints? Then you could have a single queue holding all of them. I think this would allow you to write this code much more compactly and 'functionally'. You could even have a struct (idx, hash, tag) that is shared in both cases. Something like the following (which is hand-wavy and illustrative only):
```
let payload = Whatever { idx, hash, tag}

let to_push = match path_kind {
  Concrete => Enforcement::Unconditional(payload),
  MaybeConcrete(x) => Enforcement::Conditional { path: concrete_path.clone(), payload }
};

hash_slots.hash2_stacks.the_only_stack(to_push);
```

Reply:
To simplify code-sharing between the various arities, the assignment to `to_push` above could have its logic extracted and become something like
```
let to_push = Enforcement::new(path_kind, payload);
```

Reply:
etc. you get the point. I won't work it to death, but I think there's a natural endpoint that will make all this *much* clearer — and hopefully in a way that cascades throughoutl.

Reply:
We're using implications everywhere, now that they're cheap. The need for two kinds of enforcements is now gone